File: tools/scribble/stitch.py
"""Cut pen lifts by reordering and joining strokes.

The greedy walk produces thousands of short chains because it keeps stranding
itself in paper it has already paid off. Each chain is a pen lift, and a lift
costs about two seconds of machine time regardless of how long the stroke
either side of it is, so the chain count - not the ink - sets the plot time.

Two passes, in order:

  order()  greedy nearest-neighbour over chain endpoints, either end allowed,
           so the pen finishes each chain next to the start of the following
           one. Does not change the drawing at all; it only shortens the
           pen-up travel and sets up the second pass.

  join()   merge consecutive chains whose ends are now close enough that
           drawing the gap costs less than lifting over it. This does change
           the drawing: it adds ink the image did not ask for, which is why
           the threshold wants to stay near a nib width or two.
"""
import numpy as np


def order(polylines):
    """Greedy nearest-neighbour tour of the chains, reversing where it helps."""
    if len(polylines) < 3:
        return list(polylines)
    starts = np.array([pl[0] for pl in polylines])
    ends = np.array([pl[-1] for pl in polylines])
    n = len(polylines)
    used = np.zeros(n, dtype=bool)

    out = [polylines[0]]
    used[0] = True
    p = ends[0]
    for _ in range(n - 1):
        ds = np.where(used, np.inf, ((starts - p) ** 2).sum(axis=1))
        de = np.where(used, np.inf, ((ends - p) ** 2).sum(axis=1))
        i_s, i_e = int(np.argmin(ds)), int(np.argmin(de))
        if de[i_e] < ds[i_s]:
            out.append(polylines[i_e][::-1])
            p = starts[i_e]
            used[i_e] = True
        else:
            out.append(polylines[i_s])
            p = ends[i_s]
            used[i_s] = True
    return out


def join(polylines, max_gap):
    """Merge consecutive chains separated by no more than max_gap."""
    if not polylines:
        return []
    out = [polylines[0]]
    for pl in polylines[1:]:
        gap = float(np.hypot(*(pl[0] - out[-1][-1])))
        if gap <= max_gap:
            out[-1] = np.concatenate([out[-1], pl], axis=0)
        else:
            out.append(pl)
    return out


def tidy(polylines, max_gap=0.0):
    """order(), then join() if a gap budget is given."""
    out = order(polylines)
    return join(out, max_gap) if max_gap > 0 else out


def travel(polylines):
    """(drawn length, pen-up length) - what the reorder is actually buying."""
    drawn = up = 0.0
    prev = None
    for pl in polylines:
        if len(pl) > 1:
            drawn += float(np.hypot(*np.diff(pl, axis=0).T).sum())
        if prev is not None:
            up += float(np.hypot(*(pl[0] - prev)))
        prev = pl[-1]
    return drawn, up

File: tsc/src/scribble/stitch.ts
// Cutting the pen lifts out of a scribble.
//
// The greedy walk produces thousands of short chains, because it keeps
// stranding itself in paper it has already paid off. Each chain is a pen lift,
// and a lift costs about four seconds of machine time - a pen-down, a pen-up,
// and the travel between - regardless of how long the strokes either side of it
// are. So the chain count, not the ink, sets the plot time: a horse drawn in
// 1,149 chains spends over an hour lifting.
//
// Two passes, in this order, and they are not the same kind of change:
//
//   order()  reorders and reverses chains so the pen finishes each one near the
//            start of the next. Does not alter the drawing at all - the same
//            strokes in a different sequence - so it is free.
//
//   join()   merges consecutive chains whose ends are now close enough that
//            drawing the gap is cheaper than lifting over it. This DOES alter
//            the drawing: it adds ink the image never asked for, which is why
//            the threshold belongs near a nib width or two and not higher.
import { Point } from './greedy';

function distance(a: Point, b: Point): number {
    return Math.hypot(a.x - b.x, a.y - b.y);
}

/**
 * Greedy nearest-neighbour tour of the chains, reversing one where that is the
 * closer end.
 *
 * O(n^2) over chains, which is fine at the few thousand a walk produces and is
 * the same shape as the optimiser the renderer already runs over infill.
 */
export function orderChains(chains: Point[][]): Point[][] {
    if (chains.length < 3) return chains.slice();

    const used = new Array<boolean>(chains.length).fill(false);
    const ordered: Point[][] = [chains[0]];
    used[0] = true;
    let at = chains[0][chains[0].length - 1];

    for (let placed = 1; placed < chains.length; placed++) {
        let bestIndex = -1;
        let bestDistance = Infinity;
        let bestReversed = false;

        for (let i = 0; i < chains.length; i++) {
            if (used[i]) continue;
            const chain = chains[i];
            const toStart = distance(at, chain[0]);
            if (toStart < bestDistance) {
                bestDistance = toStart;
                bestIndex = i;
                bestReversed = false;
            }
            const toEnd = distance(at, chain[chain.length - 1]);
            if (toEnd < bestDistance) {
                bestDistance = toEnd;
                bestIndex = i;
                bestReversed = true;
            }
        }

        const chain = bestReversed ? chains[bestIndex].slice().reverse() : chains[bestIndex];
        used[bestIndex] = true;
        ordered.push(chain);
        at = chain[chain.length - 1];
    }

    return ordered;
}

/**
 * Merges consecutive chains separated by no more than `maxGapMm`.
 *
 * Run this after orderChains, never before: on the order the walk produced
 * them, consecutive chains are wherever the walk happened to restart, and
 * almost none of them are adjacent.
 */
export function joinChains(chains: Point[][], maxGapMm: number): Point[][] {
    if (!(maxGapMm > 0) || chains.length === 0) return chains.slice();

    const joined: Point[][] = [chains[0].slice()];
    for (let i = 1; i < chains.length; i++) {
        const previous = joined[joined.length - 1];
        const gap = distance(previous[previous.length - 1], chains[i][0]);
        if (gap <= maxGapMm) {
            previous.push(...chains[i]);
        } else {
            joined.push(chains[i].slice());
        }
    }
    return joined;
}

export type StitchSummary = {
    chains: number;
    drawnMm: number;
    travelMm: number;
};

/** What a set of chains costs: ink drawn, and pen-up travel between them. */
export function measureChains(chains: Point[][]): StitchSummary {
    let drawn = 0;
    let travel = 0;
    let previousEnd: Point | undefined;

    for (const chain of chains) {
        for (let i = 1; i < chain.length; i++) {
            drawn += distance(chain[i - 1], chain[i]);
        }
        if (previousEnd) travel += distance(previousEnd, chain[0]);
        previousEnd = chain[chain.length - 1];
    }

    return { chains: chains.length, drawnMm: drawn, travelMm: travel };
}

/** order(), then join() when a gap budget is given. */
export function stitch(chains: Point[][], maxGapMm = 0): Point[][] {
    const ordered = orderChains(chains);
    return maxGapMm > 0 ? joinChains(ordered, maxGapMm) : ordered;
}
